From d0b60f99e24ff92ebb971f48bfb45ef936cf6442 Mon Sep 17 00:00:00 2001
From: Tomas Glozar <tglozar@gmail.com>
Date: Tue, 21 Nov 2023 16:40:42 +0100
Subject: [PATCH 15/94] ia64: set nid of all reserved memblocks to 0 at setup

Kernel panic was reported on HP Integrity rx6600 servers after commit
61167ad5fecde ("mm: pass nid to reserve_bootmem_region()"):

Unable to handle kernel paging request at virtual address
0000000000007980
swapper[0]: Oops 11012296146944 [1]
Modules linked in:
CPU: 0 PID: 0 Comm: swapper Not tainted
6.4.0-rc4-61167ad5fecdeaa037f3df1ba354dddd5f66a1ed-ia64 #1
psr : 00001210084a2010 ifs : 8000000000000186 ip : Not tainted
6.4.0-rc4-61167ad5fecdeaa037f3df1ba354dddd5f66a1ed-ia64
ip is at reserve_bootmem_region (mm/mm_init.c:651 mm/mm_init.c:701
mm/mm_init.c:748)

This is caused by reserved memory blocks being referenced before correct
nid is set in memmap_init_reserved_pages().

Fixes: 61167ad5fecde ("mm: pass nid to reserve_bootmem_region()")
---
 arch/ia64/kernel/setup.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/ia64/kernel/setup.c b/arch/ia64/kernel/setup.c
index 4faea2d2cf07..f6d6a3b5d1f0 100644
--- a/arch/ia64/kernel/setup.c
+++ b/arch/ia64/kernel/setup.c
@@ -418,6 +418,8 @@ reserve_memory (void)
 
 		memblock_reserve(addr, size);
 	}
+
+	memblock_set_node(0, PHYS_ADDR_MAX, &memblock.reserved, 0);
 }
 
 /**
-- 
2.25.1

